warehouse/query-core/src/main/java/datawave/next/scanner/DocumentScanner.java

```diff
@@ -0,0 +1,127 @@
+import datawave.query.tables.BatchScannerSession;
+import datawave.query.tables.async.Scan;
+import datawave.query.tables.async.event.VisitorFunction;
+
+/**
+ * An alternate to the {@link BatchScannerSession}.
+ * <p>
+ * The {@link Scan} is replaced by the concept of a tablet worker. This worker operates two scanners in parallel, one scanner finds candidate documents and the
+ * other scanner submits document-range queries.
+ */
+public class DocumentScanner implements Iterator<Result>, Closeable {
+
+    private static final Logger log = LoggerFactory.getLogger(DocumentScanner.class);
+
+    protected Result result;
+
+    private final ArrayBlockingQueue<Result> results;
+
+    // special executor pool for our consumers
+    private final ExecutorService consumerExecutorPool = Executors.newFixedThreadPool(2);
+
+    // fetches document ids
+    private final ExecutorService docIdExecutorPool;
+
+    // fetches and evaluates document candidates
+    private final ExecutorService documentExecutorPool;
+
+    protected final DocumentScannerConfig config;
+
+    private final Iterator<QueryData> queryDataIterator;
+
+    /**
+     * Default constructor, will likely need to swap this out for a config object constructor
+     *
+     * @param config
+     *            the {@link DocumentScannerConfig}
+     * @param queryDataIterator
+     *            the iterator of {@link QueryData}
+     */
+    public DocumentScanner(DocumentScannerConfig config, Iterator<QueryData> queryDataIterator) {
+        this.config = config;
+        this.queryDataIterator = queryDataIterator;
+        this.results = this.config.getResults();
+
+        this.docIdExecutorPool = this.config.getDocIdExecutorPool();
+        this.documentExecutorPool = this.config.getDocumentExecutorPool();
+    }
+
+    public void start() {
+        // takes query data and either submits fi scans or pushes document scans directly to the doc id queue
+        QueryDataConsumer queryDataConsumer = new QueryDataConsumer(config, queryDataIterator);
+        consumerExecutorPool.execute(queryDataConsumer);
+
+        // a document id consumer creates document range scans which push results onto the results queue
+        DocumentIdConsumer docIdConsumer = new DocumentIdConsumer(config);
+        consumerExecutorPool.execute(docIdConsumer);
+    }
+
+    public void setVisitorFunction(VisitorFunction visitorFunction) {
+        this.config.setVisitorFunction(visitorFunction);
+    }
+
+    public Iterator<Result> iterator() {
+        hasNext();
+        return this;
+    }
+
+    @Override
+    public boolean hasNext() {
+        while ((config.getDocumentIdConsumerExecuting().get() || !config.getDocIdQueue().isEmpty() || config.getNumFiScans().get() > 0
+                        || config.getNumDocScans().get() > 0 || !config.getResults().isEmpty()) && result == null) {
+            try {
+                result = results.poll(1, TimeUnit.MILLISECONDS);
```

Comment:
fast polling is potentially painful for the cpu... self promotion, the ProducerConsumerBuffer might be a better fit here. 
https://github.com/NationalSecurityAgency/datawave/blob/feature/producerConsumerBuffer/warehouse/query-core/src/main/java/datawave/query/tables/ProducerConsumerBuffer.java

Reply:
or as an Iterator

https://github.com/NationalSecurityAgency/datawave/blob/feature/producerConsumerBuffer/warehouse/query-core/src/main/java/datawave/query/tables/ConsumerIterator.java

Reply:
Created https://github.com/NationalSecurityAgency/datawave/issues/2912